drivers/gpio/gpio_pca95xx.c

```diff
@@ -0,0 +1,671 @@
+#define REG_INT_STATUS_PORT1		0x4D
+#define REG_OUTPUT_PORT_CONF		0x4F
+
+/* Driver flags */
+#define PCA_HAS_PUD			BIT(0)
+
+/**
+ * @brief Check to see if a I2C master is identified for communication.
+ *
+ * @param dev Device struct.
+ * @return 1 if I2C master is identified, 0 if not.
+ */
+static inline int has_i2c_master(struct device *dev)
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	struct device * const i2c_master = drv_data->i2c_master;
+
+	if (i2c_master) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+/**
+ * @brief Read both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, read the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the PCA95XX.
+ * @param reg Register to read (the PORT0 of the pair of registers).
+ * @param buf Buffer to read data into.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int read_port_regs(struct device *dev, u8_t reg,
+			   union gpio_pca95xx_port_data *buf)
+{
+	const struct gpio_pca95xx_config * const config =
+		dev->config->config_info;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	struct device * const i2c_master = drv_data->i2c_master;
+	u16_t i2c_addr = config->i2c_slave_addr;
+	int ret;
+
+	ret = i2c_burst_read(i2c_master, i2c_addr, reg, buf->byte, 2);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error reading register 0x%X (%d)",
+			i2c_addr, reg, ret);
+		goto error;
+	}
+
+	LOG_DBG("PCA95XX[0x%X]: Read: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		i2c_addr, reg, buf->byte[0], (reg + 1), buf->byte[1]);
+
+error:
+	return ret;
+}
+
+/**
+ * @brief Write both port 0 and port 1 registers of certain register function.
+ *
+ * Given the register in reg, write the pair of port 0 and port 1.
+ *
+ * @param dev Device struct of the PCA95XX.
+ * @param reg Register to write into (the PORT0 of the pair of registers).
+ * @param buf Buffer to write data from.
+ *
+ * @return 0 if successful, failed otherwise.
+ */
+static int write_port_regs(struct device *dev, u8_t reg,
+			    union gpio_pca95xx_port_data *buf)
+{
+	const struct gpio_pca95xx_config * const config =
+		dev->config->config_info;
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	struct device * const i2c_master = drv_data->i2c_master;
+	u16_t i2c_addr = config->i2c_slave_addr;
+	int ret;
+
+	LOG_DBG("PCA95XX[0x%X]: Write: REG[0x%X] = 0x%X, REG[0x%X] = "
+		"0x%X", i2c_addr, reg, buf->byte[0], (reg + 1),
+		buf->byte[1]);
+
+	ret = i2c_reg_write_byte(i2c_master, i2c_addr, reg, buf->byte[0]);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error writing to register 0x%X "
+			"(%d)", i2c_addr, reg, ret);
+	}
+	ret = i2c_reg_write_byte(i2c_master, i2c_addr, reg+1, buf->byte[1]);
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error writing to register 0x%X "
+				"(%d)", i2c_addr, reg, ret);
+	}
```

Comment:
Performing two independent i2c bus write transactions is a bit expensive. It would be better to use `i2c_burst_write`. According to the datasheet it's supported by the chip. We only need to be careful that `buf` payload is sent correctly on big-endian and little-endian MCUs.